cpp/ql/src/Likely Bugs/Likely Typos/illDefinedForLoop.ql

```diff
@@ -0,0 +1,102 @@
+/**
+ * @name Ill-defined for loop
+ * @description A for-loop iteration expressions goes backward with respect of the initialization statement and condition expression.
+ * @id cpp/ill-defined-for-loop
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @tags external/microsoft/6293
+ * @msrc.severity important
+ */
+import cpp
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+predicate illDefinedDecrForStmt( ForStmt forstmt, Variable v, Expr initialCondition, Expr terminalCondition ) { 
+  v.getAnAssignedValue() = initialCondition 
+  and
+  exists( 
+    RelationalOperation rel, Expr e |
+    rel = forstmt.getCondition() |
+    e = rel.getGreaterOperand()
+    and v.getAnAccess() = rel.getLesserOperand()
+    and terminalCondition = e
+  )
+  and
+  (
+    exists( 
+      PostfixDecrExpr pdec |
+      pdec = forstmt.getUpdate().(PostfixDecrExpr)
+      and pdec.getAnOperand() = v.getAnAccess()
+    ) or 
+    exists( 
+      PrefixDecrExpr pdec |
+      pdec = forstmt.getUpdate().(PrefixDecrExpr)
+      and pdec.getAnOperand() = v.getAnAccess()
+    )
+  )
+  and
+  upperBound(initialCondition) < lowerBound(terminalCondition)
```

Comment:
So we flag:
```
for(int i = 0; i < 100; i--) // initialcondition < terminalcondition
```
but is there a reason not to also flag:
```
for(int i = 200; i < 100; i--) // initialcondition > terminalcondition
```

Reply:
The second ill-defined loop can be found by checking if ( forstmt.conditionAlwaysFalse() or forstmt.conditionAlwaysTrue() ), I actually considered adding this check to this rule, but since C6293 was not looking for this condition, I thought about writing a separate rule. If you want I can put both checks.

Reply:
I'd prefer that "loop never taken" be a separate rule. This rule is about the relationship between the condition and the update ; "loop never taken" is about the relationship between the initialization and the condition.

Reply:
I found an interesting corner case on C6293:

> Not found by PREfast as we were discussing
`for (unsigned int i = 200; i < 100; i--)  

> Found by PREfast as C6293, but the text is incorrect ("ill defined loop: counts down from minimum")
`signed int min = 100;
signed int max = 200;
for (unsigned int i = max; i < min; i--)`

I can make a slight change to the code so we can detect this bug and give a more useful error message:
"Ill-defined for-loop: a loop using variable \"" + v + "\" counts " 
      + forLoopdirection(isIncr) + " from a value ("+ initialCondition +"), but the terminal condition is always false."

Reply:
BTW. Testing this change I found that Semmle doesn't seem to detect the following loop as .conditionAlwaysTrue():
`for (unsigned char i = 200; i >= 0; i++)`
The query would still detect the mismatch iterator, but the error message will be off for this one (I will add a comment on the test case)

Reply:
@jbj do you know why `conditionAlwaysTrue()` doesn't apply here?

Reply:
I agree it would be sensible to add "unsigned >= 0" to `conditionAlwaysTrue`.

The example (`for (unsigned char i = 200; i >= 0; i++)`) should hopefully be flagged by [UnsignedGEZero.ql](https://github.com/Semmle/ql/blob/a10c3bcffb004cd5085bc04f688018d656cfe38f/cpp/ql/src/Likely%20Bugs/Arithmetic/UnsignedGEZero.ql).

More sophisticated cases can be detected with [reachablePointlessComparison](https://github.com/Semmle/ql/blob/a10c3bcffb004cd5085bc04f688018d656cfe38f/cpp/ql/src/semmle/code/cpp/rangeanalysis/PointlessComparison.qll#L154).